### ISSUE TITLE
Support disabling src/dst checks on AWS ENIs.

### ISSUE THREAD

**Issue by Buffer0x7cd** [cilium/cilium#14243]
## Proposal / RFE

Hey everyone, While testing DSR on AWS we discovered that src/dst checks needed to be disabled in AWS ec2 instances for DSR to work (more info https://github.com/cilium/cilium/issues/13600). While the above solution worked, Internally in our team we had a discussion on how to disable this consistently across Auto scaling groups instances as AWS does not have an API to enable/disable DSR on a group of instances (src/dst API is only limited to an instance). To achieve this we thought about two solutions.     

1. Disable SRC/DSR check during using user data. In this method, each instance will be responsible for disabling its own src/dst check.    
2. Use some external component that has the knowledge of all instances and then can go and disable the check on those instances. In this case, one potential solution was to add this functionality in the Cilium CNI operator ( As it's already responsible for manipulating the ENI). Although one potential drawback of this is that, This setup might require some coordination between operator and cilium agents ( As we don't want agents to start working before the src/dst check is disabled)  otherwise it can break the networking setup in the cluster.
Let us know your thoughts on the above two approaches (or if there are some other ones that you think might be a better way). We are happy to make the contributions if the 2nd approach seems a better fit

Relevant discussion thread on slack: https://cilium.slack.com/archives/C1MATJ5U5/p1606914131477800

**Comment by errordeveloper** [comment]
Since this is done per-ENI, it's best to consider making it part of `ciliim-operator`. There is already an implicit dependency on the operator in the agent, since IPs need to be allocated prior to agent starting up. On node reboot ENI settings should be persisted, since it's in the API and not a local state. However, currently operator manages secondary ENIs, but it will need make update  to primary ENIs for this, since that's where `NodePort` service traffic is routed.

NB: at present the agent has access to the same APIs as the operator, so it could theoretically make these API calls, however that is only due to the fact that instance roles are configured. In the future instance roles will get reduced and a specific role will be assigned to the operator and that should be the only component that intends to manage ENIs. The docs don't show this yet, but some folks are setting up their clusters this way already (see #13270).

**Comment by Buffer0x7cd** [comment]
Based on the above-mentioned points, It seems like integrating this in Cilium Operator makes sense. On the other hand, if the Eksctl or AWS API team are ready to make this an option for ASG then this requirement can be offloaded to the cluster provisioning tool and the Operator will not require to handle this.

**Comment by errordeveloper** [comment]
It maybe plausible to still implement a check in the operator, so we can detect misconfiguration and warn the use.

**Comment by Buffer0x7cd** [comment]
Agree. It would be really helpful if there is some kind of check also, Otherwise tracking down the issue takes lots of effort.

**Comment by ungureanuvladvictor** [comment]
@Buffer0x7cd / @errordeveloper -- I've put some thought into this and I think we should put this functionality into the operator given that is the central spot that handles AWS API comms. Until now the operator did not do any write modifications to the primary interface of the instance (it used it just for read operations to infer things like subnet, securitygroups, etc) -- which brings the question of managing that interface -- should Cilium take it under its purview.

I think the best course of action is to add a separate flag to the operator which defaults to true if the DSR is enabled. This gives users to not manage this setting via the operator if they do it somehow else (scripts at instance boot-up, etc).

I'm not that familiar with how DSR works but do we need to support this for the interfaces that the operator creates?

As to the tactical changes we can just add a new parameter to https://github.com/cilium/cilium/blob/master/pkg/aws/eni/instances.go#L41 and pipe that through all the way to the ec2 client internally.

Another question: I assume this is not working on Azure as well? I see that we can achieve the same result via https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-network-interface#enable-or-disable-ip-forwarding -- just need to pipe that through the Azure client inside the operator.

**Comment by Buffer0x7cd** [comment]
Regarding 

> I'm not that familiar with how DSR works but do we need to support this for the interfaces that the operator creates?     

In most cases, Only the primary interface is used for the NodePort Services  (Unless the user has mentioned multiple interfaces for NodePort) due to this we only need this to be enabled for the primary interface. @brb Can you please confirm this?

> Another question: I assume this is not working on Azure as well?       

I haven't tested this on azure yet, So don’t have much information on it.

**Comment by brb** [comment]
> In most cases, Only the primary interface is used for the NodePort Services (Unless the user has mentioned multiple interfaces for NodePort) due to this we only need this to be enabled for the primary interface. @brb Can you please confirm this?

Why does this matter in the context of this issue? The src/dst check is per node, i.e. not per iface.

**Comment by ungureanuvladvictor** [comment]
@brb it's actually per ENI and not per node -- https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html#modify-source-dest-check.

**Comment by brb** [comment]
> it's actually per ENI and not per node

Oh, I see. Thanks for the clarification.

> In most cases, Only the primary interface is used for the NodePort Services (Unless the user has mentioned multiple interfaces for NodePort) due to this we only need this to be enabled for the primary interface. @brb Can you please confirm this?

I'm not very familiar with AWS, but it might be enough to disable the check for an interface which is used to communicate between k8s Nodes (aka `--direct-routing-device`). IIRC, the drop happened after the intermediate node had forwarded the request which was `| src=client_from_outside_IP | dst=pod_IP |`. However, the assumption needs validation, as I'm not sure whether the reply which is ` | src=service_IP | dst=client_from_outside_IP |` won't get dropped by the VPC if the outgoing interface is not the same as the one used for direct routing.

**Comment by ungureanuvladvictor** [comment]
@brb -- after reading AWS docs I was able to confirm that this happens just for packets that are arriving on the machine, this check is not taken into consideration where the source is different.

**Comment by brb** [comment]
@ungureanuvladvictor Thanks for the update. Then the check needs to be disabled only for this:

>  <..> to disable the check for an interface which is used to communicate between k8s Nodes (aka --direct-routing-device).

**Comment by lht** [comment]
I've been looking at implement this.

@brb  could you suggest how to pass the direct routing device's ENI id to the operator? From my understanding, the direct routing device detection is done by the agent and the operator has no knowledge of how to map device name to ENI interface.

An alternative is letting the operator to disable the `SourceDestCheck` flag for all the attached ENIs. But as being pointed out, flipping the flag on the direct routing device is suffice. 

BTW, I've tested and can confirm that when `SourceDestCheck` is enabled, outgoing packets would be dropped, if the source address doesn't match any address of the instance.

<details>
  <summary>Test details</summary>

This script (steal from [here](https://inc0x0.com/icmp-ip-packets-ping-manually-create-and-send-icmp-ip-packets/)) sends an ICMP echo request to 10.1.159.42 host A with spoofing source address 8.8.8.8. Only when sourceDestCheck on host A is disabled can the packet be received by the destination.

```python
import socket
import time

s = socket.socket(socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_TCP)
s.setsockopt(socket.IPPROTO_IP, socket.IP_HDRINCL, 1)

ip_header = b'\x45\x00\x00\x1c' # Version, IHL, Type of Service | Total Length
ip_header += b'\xab\xcd\x00\x00' # Identification | Flags, Fragment Offset
ip_header += b'\x40\x01\x6b\xd8' # TTL, Protocol | Header Checksum
ip_header += b'\x08\x08\x08\x08' # Source Address 8.8.8.8
ip_header += b'\x0a\x01\x9f\x42' # Destination Address 10.1.159.42

icmp_header = b'\x08\x00\xe5\xca' # Type of message, Code | Checksum
icmp_header += b'\x12\x34\x00\x01' # Identifier | Sequence Number

packet = ip_header + icmp_header
s.sendto(packet, ('8.8.8.8', 0))
```

Disable sourceDestCheck on host A and run tcpdump on destination,
```
root@ip-10-1-159-42:~# tcpdump -n -i any icmp and host 8.8.8.8
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked v1), capture size 262144 bytes
13:53:03.962753 IP 8.8.8.8 > 10.1.159.42: ICMP echo request, id 4660, seq 1, length 8
13:53:03.962814 IP 10.1.159.42 > 8.8.8.8: ICMP echo reply, id 4660, seq 1, length 8
```

 
</details>

**Comment by brb** [comment]
> could you suggest how to pass the direct routing device's ENI id to the operator

We are trying to get rid of the "direct routing device" notion - https://github.com/cilium/cilium/issues/17441.

**Comment by ozankabak** [comment]
I think this may cause conformance tests to fail when using native networking with ENIs
